### ISSUE TITLE
Kong ingress controller failed to fetch secrets for kong consumers if restarted (either due to some crash or planned)

### ISSUE THREAD

**Issue by rarora-21** [Kong/kubernetes-ingress-controller#5784]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

If one of KIC Pod restarts due to some issue or in planned way, it is unable to fetch secret and thus results in resource processing failed for Kong consumers.

In order to solve this: we have to delete the Kong consumers (it shows programmed false) and re-sync them into K8s cluster.

We have single deployment for both proxy and Ingress controller with DB-LESS mode

### Expected Behavior

Kong Ingress controller should able to process the existing consumer and secret

### Steps To Reproduce

```markdown
1. Run the deployment with proxy v3.4.x and KIC: 2.12.3 (lets says with replica config of count 3)
2. Create Kong consumers having credentials obtaining from K8s secret
3. Delete one of the pods and wait for restart of the containers. 
4. Once the new pod is running, check the Kong ingress controller logs  which still shows the resource processing error to process the Kong consumer and consumer doesn't work as well

FYI: Kong Consumer is created in corresponding to JWT Kong plugin
```

### Kong Ingress Controller version

```shell
KIC:2.12.3 & Proxy: 3.4.x
```

### Kubernetes version

```shell
v1.27.5
```

### Acceptance Criteria

- [ ] `KongConsumers` are correctly loaded when KIC restarted, not affected by latency of secrets - Other resources reference relationship loaded to cache.

### Anything else?

![image](https://github.com/Kong/kubernetes-ingress-controller/assets/96419636/8e299425-e03f-4d78-9c70-6b569ec6f9fb)

Above image shows error after restarting pod for one of the consumers. However, at the same time...secret and consumers are present in the cluster.

Meanwhile, deleting and recreating the Kong consumers resolve the issue to process them successfully by restarted Kong controller container

**Comment by rarora-21** [comment]
Also, to update I am using latest CRDS as stated below

    **controller-gen.kubebuilder.io/version: v0.13.0**

**Comment by oleksandrs-adorama** [comment]
looks like we have same issue 
https://github.com/Kong/kubernetes-ingress-controller/issues/5710

**Comment by rainest** [comment]
Do you see any difference when using [the latest nightly](https://hub.docker.com/r/kong/nightly-ingress-controller/tags?page=1&ordering=last_updated) when your Secrets have the `konghq.com/credential` label (instead of the `kongCredType` field)?

Other than the known (if not fully explained) issues when using older CRDs, I'm not sure what would cause this at startup. AFAIK the other issue with Secrets not being found occurs when creating a new Secret and KongConsumer simultaneously and would show a different error than the "failed to fetch" error you're seeing.

https://github.com/Kong/kubernetes-ingress-controller/commit/6db609c996c7cb24c85273d371958a853220aade does at least simplify when we load Secrets, so it's worth a try in case something odd is happening with the older reference behavior.

**Comment by rarora-21** [comment]
I can try this but I am still using the KIC v2.12.3. Is this annotation not introduced in KIC v3.0.0?

**Comment by rarora-21** [comment]
ok this doesn't work since replacing kongCredType results in another error:

time="2024-04-09T06:03:39Z" level=error msg="resource processing failed: credential \"XXXXX\" failure: failed to provision **credential: invalid kongCredType: type 'string'** not string" GVK="configuration.konghq.com/v1, Kind=KongConsumer" name=XXXX  namespace=QA

**Comment by rainest** [comment]
The annotation requires 3.0 yes, and the simplified Secret loading behavior has not been released yet, hence why it requires a nightly image.

Currently released versions relied on tracking references between KongConsumers and Secrets to determine which Secrets to load. If touching the KongConsumer loads the Secret properly, it's likely that reference isn't being handled properly. The unreleased changes just load any Secret with a credential label, so there's no opportunity for any race condition there.

https://github.com/Kong/kubernetes-ingress-controller/issues/5175 is the other report that describes that race condition, and debug logs on your instance will probably mention the Secret name before the KongConsumer, indicating that it hadn't built the reference when it determined whether to ingest the Secret.

On 2.12 you can try increasing `--init-cache-sync-duration` (aka `CONTROLLER_INIT_CACHE_SYNC_DURATION` in the environment or `ingressController.env.init_cache_sync_duration` in the chart) to a value higher than the default (`5`) to see if that makes any difference.

Due to some limitations in how the controller generates configuration, we can't use the underlying Kubernetes controller libraries as we should. Increasing that value pauses some controller tasks at startup to give the lower level Kubernetes configuration ingest time to pull in more objects. This isn't 100% reliable (it's a workaround; we're trying to find a more reliable way of handling it), but may help.

**Comment by rarora-21** [comment]
ok but I have already tried implemented `init-cache-sync-duration` and specifically upgrade from 2.11 to 2.12.3 to use this value but seems like as you said  it's not reliable and didn't work in my case.

I am currently in process of upgrading KIC to v3.0.x.   Once it is done, I will try to use nightly image and may be at the time, I will release v3.0.x in production , Simplified Secret loading behavior may release before that.

**Comment by rarora-21** [comment]
I have tested with the nightly update image and the issue is fixed now. No more resource processing errors are seen and consumer is working correctly

**Comment by programmer04** [comment]
The user reported that he tested a nightly image and the issue doesn't occur. Can we close it? WDYT?

cc: @rainest & @randmonkey

**Comment by rainest** [comment]
To the best of my knowledge, the resolution to this will be released with 3.2. The [revised ingest method](https://github.com/Kong/kubernetes-ingress-controller/issues/5784#issuecomment-2043768283) does not exhibit the same issue. There are no plans to backport this change to 2.x at present. If we decide to do so, that will be handled in a follow-up issue.
